### PR TITLE
Feature/19 Token Based Auth

### PR DESCRIPTION
**Pull request by jeor0980**
Summary: 
- New `accounts` admin endpoint added, can only be used by authenticating with username and password.

- `instance` endpoint now uses token authentication for authenticated methods (POST, DELETE) tokens are stored in walnut db and tied to an individual user. 

To Test: 
Follow the steps in the updated `REAME.md` and make an `admin` account and a personal  `developer` account.

Test the `instance` endpoint methods
- POST
- DELETE

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,15 +27,35 @@ CU Custom Pantheon Orchestration Tool
...
   env FLASK_DEBUG=1 FLASK_APP=run.py flask run
   ```
+
+## Authentication
+
+This application assumes there is a mongoDB database named `walnut` with a collection named `accounts`.
+
+The initial user is created manually by inserting a document into `accounts`.
+
+* Password must be stored as a bcrpyt hash
```
Comment (tirazel) [inline README.md]: 'bcrypt' misspelling

File: domain/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from .instance import instance
+from .accounts import accounts
+
+DOMAIN = {
+    "instance": instance,
+    "accounts": accounts
+}
```
Comment (tirazel) [inline domain/__init__.py]: PEP8: Missing newline at end of file

File: README.md [hunk 3]
```diff
@@ -27,15 +27,35 @@ CU Custom Pantheon Orchestration Tool
...
   source env/bin/activate.fish
   pip install -r requirements.txt
   env FLASK_DEBUG=1 FLASK_APP=run.py flask run
   ```
+
+## Authentication
+
+This application assumes there is a mongoDB database named `walnut` with a collection named `accounts`.
+
+The initial user is created manually by inserting a document into `accounts`.
```
Comment (jwfuller) [inline README.md:72]: Could you provide some actual CLI commands?

File: auth.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        return account
+
+
+def add_token(documents):
+    """
+    Appends a random string encoded in base64 format to a user document
+    """
+    # TODO: You should at least make sure that the token is unique.
+    for document in documents:
+        document["token"] = secrets.token_urlsafe()
```
Comment (jwfuller) [inline auth.py:34]: Need to hash token as well

Comment (jeor0980) [reply auth.py:34]: We've assessed this is not necessary at this time.

**Review by jwfuller — APPROVED** [review]
Works for me. Used the following to create an item on the `instance` endpoint using the token

```
curl -d '{"sid":"123456789"}' -H "Content-Type: application/json" -H "Authorization: Token hbrr13l6xc4OmqPRQIdfD_YQwQkR0tHQTexgxNYBb6Q" -X POST http://127.0.0.1:5000/instance
```

**Comment by tirazel** [comment]
Following the instructions to generate a token causes the following error using Python 3.7.6:

'''
Python 3.7.6 (default, Feb 14 2020, 11:08:10)
[GCC 8.3.1 20190507 (Red Hat 8.3.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import bcrypt
>>> print(bcrypt.hashpw("password_to_be_hashed", bcrypt.gensalt()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/bcrypt/__init__.py", line 61, in hashpw
    raise TypeError("Unicode-objects must be encoded before hashing")
TypeError: Unicode-objects must be encoded before hashing
'''

Works fine using Python 2.7.16, assuming the bcrypt module has been installed one way or another.

**Comment by tirazel** [comment]
When creating the accounts collection using mongoshell 4.2.3, accounts looks like it needs to be in quotes.
'''
> use walnut
switched to db walnut
> db.createCollection(accounts)
2020-02-14T23:07:37.120-0700 E  QUERY    [js] uncaught exception: ReferenceError: accounts is not defined :
@(shell):1:1
> db.createCollection('accounts')
{ "ok" : 1 }
>
'''

**Comment by tirazel** [comment]
When attempting to create a developer account, after creating the accounts collection and setting up the admin account, I am getting a large traceback ending in 
```
  File "/home/tirazel/Projects/walnut/auth.py", line 13, in check_auth
    bcrypt.hashpw(password, account['password']) == account['password']
ValueError: Invalid salt
```

Mongo:

```
> use walnut
switched to db walnut
> db.accounts.find()
{ "_id" : ObjectId("5e478e86542622b3190a7c1b"), "username" : "admin", "password" : "$2b$12$SbPujcjV0erz9oLLJnvuV.TbXYgfsbojhy/CX6VP9VeqV3uHfDsxi", "role" : "admin" }
>
```

**Comment by tirazel** [comment]
This was being caused by a version mismatch between the one I used to encrypt the admin password ('bcrypt') and the one that is part of walnut ('py-bcrypt').  Hashes produced by the version currently used in walnut should begin with '$2a$'.

**Review by tirazel — APPROVED** [review]
Appears to work fine.  My issues were being caused due to using a python outside of the venv with mismatched bcrypt implementations.

**Pull request merged**